rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+	// to connect to the cluster. Used ony in Azure.
+	string PublicProxyAddr = 5;
+}
+
+// AWSSSM provides options to use when executing SSM documents
+message AWSSSM {
+	// DocumentName is the name of the document to use when executing an
+	// SSM command
+	string DocumentName = 1;
+}
+
+// JoinParams configures the parameters for Simplified Node Joining.
+message DiscoveryJoinParams {
+  string TokenName = 1;
+	string Method = 2;
+	AzureJoinParams Azure = 3;
+}
+
+// AzureJoinParams configures the parameters specific to the Azure join method.
+message AzureJoinParams {
+  string ClientID = 1;
+}
+
+// AzureMatcher matches Azure resources.
+// It defines which resource types, filters and some configuration params.
+message AzureMatcher {
+	// Subscriptions are Azure subscriptions to query for resources.
+	repeated string Subscriptions = 1;
+	// ResourceGroups are Azure resource groups to query for resources.
+	repeated string ResourceGroups = 2;
+	// Types are Azure types to match: "mysql", "postgres", "aks", "vm"
+	repeated string Types = 3;
+	// Regions are Azure locations to match for databases.
+	repeated string Regions = 4;
+	// ResourceTags are Azure tags on resources to match.
+  wrappers.LabelValues ResourceTags = 5;
+	// InstallParams sets the join method when installing on
+	// discovered Azure nodes.
+	MatcherInstallParams InstallParams = 6;
+}
+
+// GCPMatcher matches GCP resources.
+message GCPMatcher {
+	// Types are GKE resource types to match: "gke".
+	repeated string Types = 1;
+	// Locations are GKE locations to search resources for.
+	repeated string Locations = 2;
+	// Tags are GCP labels to match.
+  wrappers.LabelValues Tags = 3;
+	// ProjectIDs are the GCP project ID where the resources are deployed.
+	repeated string ProjectIDs = 4;
+}
+
+```
+
+### Backward Compatibility
+
+#### `discovery_service.selector`
+If the config has this field but Teleport version is no aware of that, the user must remove it before they can start the Discovery Service.
+
+#### Watcher on `DiscoveryConfig`
+If the `discovery_service` tries to set up a watcher over `DiscoveryConfig` but Teleport Auth/Proxy is not yet aware of this resource, then `discovery_service` will log a warning but continue with the static configuration.
+
+### UX
+
+#### `discovery_service` new service configuration properties
+Users need to add the `selector` property and then a list of label matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  selector:
+    labelA: value1 # matches against a single value
+    labelB: [value2, value3] # matches against multiple values
+    labelC: "*" # matches on all values
+```
+
+#### WebAPI: manage `DiscoveryConfig` resource
+The following endpoints must be create to be used by WebAPI:
+```
+Methods:
+GET .../discoveryconfig
+GET .../discoveryconfig/:id
+POST .../discoveryconfig
+PUT .../discoveryconfig/:id
+DELETE .../discoveryconfig/:id
```

Comment:
Don't we want to leverage gRPC ?

Reply:
Yes!
Sorry, I'll add the gRPC methods right away

Reply:
Done :+1: